Custom train tracks

qaisjp opened the pull request:
This PR closes #388

Cazomino05 committed:
Add custom train tracks
(cherry picked from commit 45d1a266e8f3e50d9605610a32210a9a9a6e145a)

Cazomino05 committed:
Add missing files and fixed compile errors
(cherry picked from commit e2c5fcb83fa7df633f040dc6b0c048e2cd72bbb3)

Commit:
Small custom train tracks fix
(cherry picked from commit 76248682a3fdf8bf24c5d7e73f2bac4536f93069)

jushar committed:
Fix some custom train track crashes
(cherry picked from commit 92a1f4290625b07b651166cd333d3c43dad7ab0f)

qaisjp committed:
Fix build and merges for CTT

qaisjp committed:
Fix crash when creating a new track

qaisjp committed:
Fix FindClosestRailTrackNode to use CTrainTrackManager

qaisjp committed:
clean up a lot of the CTT api

includes code for getDefaultTrack, but it doesn't return a recognisable
train-track element yet...

qaisjp committed:
allow ints for get/setTrainTrack

qaisjp committed:
rework createTrainTrack function

qaisjp committed:
remove set node position/count

qaisjp committed:
improve clientside api

qaisjp committed:
move ResetTracks() call

qaisjp committed:
tweak headers

qaisjp committed:
Revert "move ResetTracks() call"

This reverts commit 5cc1de45c55ce738a48e366b62df8b312f7df1b1.

jushar committed:
Merge branch 'master' into feature/custom-train-tracks

jushar committed:
Move new files

jushar committed:
Remove duplicate track nodes

jushar committed:
Move train track nodes into shared file

jushar committed:
Refactor CTrainTrackManagerSA

jushar committed:
Refactor CTrainTrackSA

jushar committed:
Implement missing functionality in GameSA module

jushar committed:
Implement CTrainTrackManager::Reset

jushar committed:
Add required changes to DM module

jushar committed:
Refactor server CTrainTrack*

jushar committed:
Implement all left stuff
- Add CClientTrainTrackManager
- Remove track IDs from the server entirely
- Use track IDs only in internal Game SA logic

This commit breaks network compatibility which will be fixed soon though

jushar committed:
Fix a bunch of issues

jushar committed:
More fixes

jushar committed:
Merge branch 'master' into feature/custom-train-tracks

jushar committed:
Update Lua train tracks API

jushar committed:
Re-enable get-/setTrainTrack

jushar committed:
Linux compile fixes

jushar committed:
Fix linkLastNodes==true and document some variables

jushar committed:
Merge branch 'master' into feature/custom-train-tracks

jushar committed:
Fix issues when destroying a track

jushar committed:
Accept a table of track nodes instead of varargs

jushar committed:
Fix linux compile errors

jushar committed:
Merge branch 'master' into feature/custom-train-tracks

jushar committed:
Use 2d length for track length calculation

jushar committed:
Merge branch 'master' into feature/custom-train-tracks

jushar committed:
Fix default track sync

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

qaisjp committed:
Tweak train track oop

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

Pirulax commented:
What's the point of adding custom train tracks when train sync doesn't even work?

jushar commented:
On the one hand sync and custom train tracks are totally different from a technical perspective. On the other hand this branch also fixes a couple of issues with train sync (especially corrects varying train tracks)

Pirulax commented:
That's fantastic :D
Keep up the good work then :)

CrosRoad95 commented:
i see there's only createTrack option, i think that i can remove it by destroyElement, but what with default tracks ?

qaisjp commented:
default tracks appear to be broken (either a recent commit from before the merge or because of the merge), see card https://github.com/multitheftauto/mtasa-blue/projects/2#card-11515327

it should be possible to delete default tracks as this allows for total conversion mods.

(at the moment it seems that default tracks are just completely broken even if you don't use the functions — spawning a train and warping to it crashes the client)

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

qaisjp committed:
Add client veh train track oopdefs

qaisjp commented:
Work on this branch will continue and discussion will be held in #388.

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

qaisjp committed:
Fix build by applying refactors from master

Applies:
- 379a2ff6e943d6f39cbf84185af5511d365f7d02
- a63e65ce06e725f70f690d733bcb7ac3b0b87282

qaisjp committed:
Fix crash when using tram

qaisjp committed:
Fix default track sync issue introduced in 5d743da34d9

Previously this line would get triggered:
- https://github.com/multitheftauto/mtasa-blue/blob/fdc13c28006816a/Shared/sdk/net/bitstream.h#L325

This happens because of:
- https://github.com/multitheftauto/mtasa-blue/blob/e11685cab4beb7958ab202261f9c9d9b4ce71e58/Client/mods/deathmatch/logic/CClientEntity.cpp#L33-L37

qaisjp committed:
Merge commit 'master' into feature/custom-train-tracks

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

qaisjp committed:
Fix line endings

qaisjp committed:
Fix IterBegin not being present

qaisjp committed:
Mute extra whitespace annoyance

qaisjp commented:
Something else is terribly fucked now, train won't even move, lol.

qaisjp committed:
Don't use asm for createMissionTrain

qaisjp committed:
Return normal numbers for getTrackNodePosition

qaisjp committed:
Prevent empty tracks from being created because it crashes

qaisjp commented on Client/game_sa/CPoolsSA.cpp:
```diff
@@ -871,28 +872,34 @@ CVehicle* CPoolsSA::AddTrain(CClientVehicle* pClientVehicle, CVector* vecPositio
+        // NOTE(qaisjp): I have no idea what the below comment is talking about.
+        // place as close to point as possible (rather than at node)? (maybe) (actually seems to have an effect on the speed, so changed from 1 to 0)
+        0);
 
```
> There is one problem: you can't set the train position of streamed out trains. Try placing two trains (`t1` and `t21`) on the same track, near each other (trainPositions `p1` and `p2`). Then disconnect, and use your server console to place them at the exact same position (set `p1 = p2`). Then reconnect (or restart the game and reconnect) and you'll find that the trains have moved.



TODO: we could set the train position here? otherwise the client's train will not match what the server requested it to be. there might be a better time for setting the train position than here, though, such as when the train streams in

qaisjp committed:
Merge branch 'master' into feature/custom-train-tracks

qaisjp committed:
Fix build

qaisjp committed:
tweak some formatting

qaisjp committed:
Use new parser for some shared things

qaisjp committed:
Remove "Track" from (Get|Set)TrackNodePosition

qaisjp committed:
Reimplement Lua API in new parser

qaisjp committed:
Tweak some defines

patrikjuvonen committed:
Merge branch 'master' into feature/custom-train-tracks
